src/OrbitQt/TrackTypeItemModel.cpp

```diff
@@ -0,0 +1,142 @@
+  CHECK(idx.column() >= 0 && idx.column() < static_cast<int>(Column::kEnd));
+
+  Track::Type track_type = known_track_types_.at(idx.row());
+  Column col = static_cast<Column>(idx.column());
+
+  switch (role) {
+    case Qt::UserRole:
+      return QVariant::fromValue(track_type);
+    case Qt::DisplayRole:
+      if (col == Column::kName) {
+        return GetTrackTypeDisplayName(track_type);
+      }
+      break;
+    case Qt::DecorationRole:
+      switch (static_cast<Column>(idx.column())) {
+        case Column::kVisibility:
+          return track_manager_->GetTrackTypeVisibility(track_type) ? visibility_on_icon_
+                                                                    : visibility_off_icon_;
+      }
+      break;
+    case Qt::CheckStateRole:
+      if (col == Column::kVisibility) {
+        return track_manager_->GetTrackTypeVisibility(track_type) ? Qt::Checked : Qt::Unchecked;
+      }
+      break;
+  }
+
+  return QVariant();
+}
+
+bool TrackTypeItemModel::setData(const QModelIndex& idx, const QVariant& value, int role) {
+  CHECK(idx.isValid());
+  CHECK(idx.model() == static_cast<const QAbstractItemModel*>(this));
+  CHECK(idx.row() >= 0 && idx.row() < static_cast<int>(known_track_types_.size()));
+  CHECK(idx.column() >= 0 && idx.column() < static_cast<int>(Column::kEnd));
+
+  Track::Type track_type = known_track_types_.at(idx.row());
+  Column col = static_cast<Column>(idx.column());
+
+  switch (role) {
+    case Qt::CheckStateRole:
+      if (col == Column::kVisibility) {
+        track_manager_->SetTrackTypeVisibility(track_type,
+                                               value.value<Qt::CheckState>() == Qt::Checked);
+        return true;
+      }
+      break;
+  }
+
+  return false;
+}
+
+QVariant TrackTypeItemModel::headerData(int section, Qt::Orientation orientation, int role) const {
+  if (orientation == Qt::Vertical) {
+    return {};
+  }
+
+  if (role == Qt::DisplayRole) {
+    switch (static_cast<Column>(section)) {
+      case Column::kVisibility:
+        return "Show";
+      case Column::kName:
+        return "Track Type";
+      case Column::kEnd:
+        UNREACHABLE();
+    }
+  }
+  return {};
+}
+
+int TrackTypeItemModel::rowCount(const QModelIndex& parent) const {
+  if (parent.isValid()) {
+    return 0;
+  }
+
+  if (track_manager_ == nullptr) {
+    return 0;
+  }
+
+  return known_track_types_.size();
+}
+
+Qt::ItemFlags TrackTypeItemModel::flags(const QModelIndex& index) const {
+  Qt::ItemFlags flags = QAbstractTableModel::flags(index);
+  if (index.column() == static_cast<int>(Column::kVisibility)) {
+    flags |= Qt::ItemIsUserCheckable;
+  }
+  return flags;
+}
+
+void TrackTypeItemModel::SetTrackManager(TrackManager* track_manager) {
+  emit layoutAboutToBeChanged();
+  track_manager_ = track_manager;
+  emit layoutChanged();
+  emit dataChanged(index(0, 0), index(rowCount() - 1, columnCount() - 1));
```

Comment:
The Qt documentation somewhat suggests that the model can use this to indicate changes to the rows, e.g. when sorting... initially I only emitted `dataChanged()`, but that resulted in the table view not showing the newly added rows.

I've tried to replace it with removeRows / insertRows, that also works well and doesn't require dealing with the correct signals, so I'll stick with that, thanks.

I don't like passing `track_manager_` in the constructor because `TrackManager` is recreated when starting a new capture, and I'm not recreating the model.

Reply:
> I don't like passing track_manager_ in the constructor because TrackManager is recreated when starting a new capture, and I'm not recreating the model.

Recreating the model might lead to simpler code and the same result. Both is fine for me though.